lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java

```diff
@@ -0,0 +1,235 @@
+    } else {
+      candidates = new TreeSet<>();
+    }
+    int size = vectors.size();
+    for (int i = 0; i < numSeed && i < size; i++) {
+      int entryPoint = random.nextInt(size);
+      candidates.add(new Neighbor(entryPoint, compare(query, vectors.vectorValue(entryPoint), searchStrategy)));
+    }
+    // set of ordinals that have been visited by search on this layer, used to avoid backtracking
+    //IntHashSet visited = new IntHashSet();
+    Set<Integer> visited = new HashSet<>();
+    // TODO: use PriorityQueue's sentinel optimization
+    Neighbors results = Neighbors.create(topK, scoreReversed);
+    for (Neighbor c :candidates) {
+      visited.add(c.node);
+      results.insertWithOverflow(c);
+    }
+    // Set the bound to the worst current result and below reject any newly-generated candidates failing
+    // to exceed this bound
+    BoundsChecker bound = BoundsChecker.create(scoreReversed);
+    bound.bound = results.top().score;
+    while (candidates.size() > 0) {
+      // get the best candidate (closest or best scoring)
+      Neighbor c = candidates.pollLast();
+      if (results.size() >= topK) {
+        if (bound.check(c.score)) {
+          break;
+        }
+      }
+      graphValues.seek(c.node);
+      int friendOrd;
+      while ((friendOrd = graphValues.nextArc()) != NO_MORE_DOCS) {
+        if (visited.contains(friendOrd)) {
+          continue;
+        }
+        visited.add(friendOrd);
+        float score = compare(query, vectors.vectorValue(friendOrd), searchStrategy);
+        if (results.size() < topK || bound.check(score) == false) {
+          Neighbor n = new Neighbor(friendOrd, score);
+          candidates.add(n);
+          results.insertWithOverflow(n);
+          bound.bound = results.top().score;
+        }
+      }
+    }
+    return results;
+  }
+
+  /**
+   * Returns the nodes connected to the given node by its outgoing arcs.
+   * @param node the node whose friends are returned
+   */
+  public int[] getNeighbors(int node) {
+    return graph.get(node).stream().mapToInt(Neighbor::node).toArray();
+  }
+
+  /** Connects two nodes symmetrically.
+   * node1 must be less than node2 and must already have been inserted to the graph */
+  void connectNodes(int node1, int node2, float score, int maxConnections) {
+    assert node1 >= 0 && node2 >= 0;
+    assert node1 < node2;
+    List<Neighbor> arcs1 = graph.get(node1);
+    assert arcs1 != null;
+    assert arcs1.isEmpty() || arcs1.get(arcs1.size() - 1).node < node2;
+    arcs1.add(new Neighbor(node2, score));
+    List<Neighbor> arcs2;
+    if (node2 < graph.size()) {
+      arcs2 = graph.get(node2);
+      assert arcs2.get(arcs2.size() - 1).node < node1;
+    } else {
+      assert node2 == graph.size();
+      arcs2 = new ArrayList<>();
+      graph.add(arcs2);
+    }
+    arcs2.add(new Neighbor(node1, score));
+
+    // ensure #arcs <= maxConnections
+    /*
+    if (maxConnections > 0) {
```

Comment:
Hmm did we decide not to impose the `maxConnections`?

Reply:
Coming in the next revision